File: file_processing/pl/filter_numbers.pl
#!/usr/bin/perl

#######################################
# Read in a file of numbers and
# pick out odds or events and compute
# their sums.
# @author: vladimir kulyukin
#######################################

use strict;
use warnings;
use List::Util qw(sum);

sub is_odd  { return $_[0] % 2 != 0; }
sub is_even { return !is_odd($_[0]); }

sub filter_lines {
    my ($file_path, $fun, $pred) = @_;
    open(IN, '<', $file_path) or die $!;
    my @rslt = ();
    while(<IN>) { push(@rslt, $fun->($_)) if ($pred->($_)); }
    return \@rslt;
}

sub pick_odds  { 
    return filter_lines($_[0], 
			sub { return int($_[0]); }, 
			\&is_odd);  
}

sub pick_evens { 
    return filter_lines($_[0], 
			sub { return int($_[0]); }, 
			\&is_even); 
}

sub sum_odds  { return sum(@{pick_odds($_[0])});  }
sub sum_evens { return sum(@{pick_evens($_[0])}); }

print $_, "\n" foreach(@{pick_odds($ARGV[0])});
print sum_odds($ARGV[0]), "\n\n";
print $_, "\n" foreach(@{pick_evens($ARGV[0])});
print sum_evens($ARGV[0]), "\n\n";

File: file_processing/py/filter_numbers.py
#!/usr/bin/python

#######################################
# Read in a file of numbers and
# pick out odds or events and compute
# their sums.
# @author: vladimir kulyukin
#######################################

import sys

def is_odd(n):  return n % 2 != 0
def is_even(n): return not is_odd(n)

def filter_lines(file_path, fun, pred):
    with open(file_path, 'r') as infile:
        return [fun(line) for line in infile if pred(line)]

def pick_odds(file_path):
    return filter_lines(file_path,
                        lambda x: int(x),
                        lambda x: is_odd(int(x)))

def pick_evens(file_path):
    return filter_lines(file_path,
                        lambda x: int(x),
                        lambda x: is_even(int(x)))

def sum_odds(file_path):  return sum(pick_odds(file_path))
def sum_evens(file_path): return sum(pick_evens(file_path))

def main(file_path):
    for n in pick_evens(file_path): print n

if __name__ == '__main__':
    main(sys.argv[1])
